### PR TITLE
DVDInterface/DVDThread: Don't store CoreTiming event IDs in savestates

### PR DESCRIPTION
**Pull request by JosJuice**
CoreTiming event types aren't guaranteed to be stable across runs of Dolphin, so they shouldn't be in savestates. Fixes https://bugs.dolphin-emu.org/issues/9466

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3759)

### CODE REVIEW

File: Source/Core/Core/HW/DVDThread.cpp [hunk 1]
```diff
@@ -47,9 +47,9 @@ static u32 s_output_address;
 static u32 s_length;
 static bool s_decrypt;
 
-// Used to notify emulated software after executing command.
-// Pointers don't work with savestates, so CoreTiming events are used instead
-static int s_callback_event_type;
+// This determines which function will be used as a callback.
+// We can't have a function pointer here, because they can't be in savestates.
+static bool s_reply_to_ios;
```
Comment (lioncash) [inline Source/Core/Core/HW/DVDThread.cpp]: It may not be particularly wise to store a bool into the savestate, since its size can vary across platforms (I realize this is done with s_decrypt, but just bringing it up as just-in-case).

Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: bools are used in savestates in other places too, like DVDInterface. If anything needs to be done about that before 5.0, it should be in a separate PR because it would touch more than the code that's related to this PR.

Comment (lioncash) [inline Source/Core/Core/HW/DVDThread.cpp]: Fair point. It can just wait until 5.0 is released, and I can just change the handling code over in one PR, which saves from doing extra work, so don't worry about it.

Comment (delroth) [inline Source/Core/Core/HW/DVDThread.cpp]: If bool's size is not stable, we should fix that in ChunkFile.

Comment (lioncash) [inline Source/Core/Core/HW/DVDThread.cpp]: Indeed. That's what I was planning to do, so it provides a transparent interface (and is less intrusive to everything else). Something like the following should suffice (note: I have not tested this, this is entirely conjecture).

```cpp
template <>
void Do(bool& value)
{
    // Make the given bool value a 32-bit value to be stable across all platforms
    // where bool may differ in terms of bit size.
    u32 stable = static_cast<u32>(value);

    DoVoid(&stable, sizeof(u32));

    if (mode == MODE_READ)
         value = stable != 0;
}
```

would you prefer a change like this to be included in 5.0, or after?

Comment (delroth) [inline Source/Core/Core/HW/DVDThread.cpp]: Included. Why not u8?

Comment (lioncash) [inline Source/Core/Core/HW/DVDThread.cpp]: That works too

Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: also why DoVoid? Do(stable) should be fine.

**Comment by phire** [comment]
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3759)*

File: Source/Core/Core/HW/DVDThread.cpp (hunk 1 above)
Comment (lioncash) [inline Source/Core/Core/HW/DVDThread.cpp]: > should be fine

Yep, that would work as well.

**Comment by RisingFog** [comment]
Anything remaining for this PR?

**Comment by JosJuice** [comment]
Not as far as I know (except for me needing to rebase it because of one of the merges that happened a few minutes ago).

**Comment by JMC47** [comment]
Rebase and LGTM.  Just make sure the savestate version is upped so that we don't get those annoying random crashes from version differences.

**Comment by JosJuice** [comment]
The savestate version is changed. You can see it by checking the changes to State.cpp.

**Comment by phire** [comment]
LGTM

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3759#-:-KF5zFpqir7WZ-HVxkRb:676834233)*

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by RisingFog** [comment]
Anything preventing this from merge?

File: Source/Core/Core/HW/DVDThread.cpp (hunk 1 above)
Comment (lioncash) [inline Source/Core/Core/HW/DVDThread.cpp]: I'll follow up with adding this when I get home from work, now that this change is merged.
